Performance Points Updates (2021) (#4768)
ppy/osu-wiki: news/2021-01-14-performance-points-updates.md

For those unfamiliar with the technical minutiae of the system, please consult [the wiki article](/wiki/Performance_points/) for an overview of the general ideas involved. 

To summarize the article very briefly (you should really read the whole article), pp encompasses several core values as follows:

- **Aim**, which considers how difficult it is to consistently hit notes back-to-back during play (affected by things like approach rate, jumps, Flashlight, Hidden, Hard Rock and more)
- **Speed**, which considers how quickly new elements are presented to the player during the course of the map (affected by things like the number of hit objects over time, BPM, Double Time and Half Time)
- **Accuracy**, which considers the player's ability to stay on time rhythmically and how difficult this is to achieve (affected by things like Hidden, Hard Rock, Flashlight and more)
- **Strain**, which considers how often and how long a player has to endure intense or technically demanding patterns during play (affected by things like streams, fast jumps and more)

> **The changes listed below apply solely to the osu! game mode.**

Author: we generally don't refer to osu! as "osu!standard" anymore, it's a holdover from Ye Olde Days

Let's get started with the details:

> ## Approach Rate 11 Changes 

Reviewer: I am fine with either approach. I did intentionally buff low AR as I felt the 1% buff was basically negligable, however I completely forgot about it during the MR. I dont believe that the change is detrimental, but since the change was not brought up in the original MR, it can totally be removed. Values from the change are still included in all changes, and very few plays / players benefit from them. As low AR is typically avoided for farming, I am confident that leaving the buff in will have a negligable effect. However, if mappers find a way to abuse this new low AR buff, it can be scaled back or adjusted or given a new formula entirely.

> ## Approach Rate 11 Changes 

Reviewer 2: After some voiced concerns I will be submitting a revert of the low AR buff later today.

> ## Approach Rate 11 Changes 

Reviewer 3: Lucky someone is paying attention, nice catch.

Previously, any map with an approach rate (hereafter referred to as AR) higher than 10.33 was affected by a buff that increased the value of the speed and aim components by a flat 20%.

Maps at or above AR 11 are often considered to be similar to the Flashlight mod in that they become considerably more difficult as a map progresses. This fact was not reflected in the previous version of the algorithm with its static 20% increase.

As suggested by [**Xexxar**](https://osu.ppy.sh/users/2773526), the following change has been applied:

- The buff for AR 11 maps now scales from +0% at 0 objects, to 26.66% at 1000 objects. The previous 20% buff is now attained at 750 objects instead.
    + Old value: 20% flat.

In essence, this nerfs the amount of pp awarded for shorter AR 11 maps, but buffs the amount gained for longer, more difficult performances.

## NoFail Changes

> Before today's changes, the performance points algorithm applied a flat 10% reduction in the maximum possible pp (not score) you could receive for a play made with the NoFail mod active.

Reviewer 4: ```suggestion
Before today's changes, the performance points algorithm applied a flat 10% reduction in the maximum possible pp (not score) you could receive for a play made with the No Fail mod active.
```

> As suggested by GitHub contributor stanriders in [this pull request](https://github.com/ppy/osu-performance/pull/127), the following change has been applied:

Reviewer 5: link to his osu profile (since you did that for xexxar)? https://osu.ppy.sh/users/7217455

Reviewer 6: Might as well use `StanR` instead of `stanriders`

Author: applied on both counts

> - NoFail's pp multiplier now scales based on the amount of misses.

Reviewer 4: ```suggestion
- No Fail's pp multiplier now scales based on the amount of misses.
```

    + New value: 1.0x base, with -0.02x applied for each miss to a maximum reduction back to 0.9x
    + Old value: 0.9x (always)

The score reduction multiplier of 0.5x remains unchanged.

> This makes sudden "pop off" performances made with the NoFail mod less punishing and more rewarding, and reduces the ability for the NoFail mod to be used by unscrupulous players to "sandbag" for rank restricted tournaments. Win-win.

Reviewer 4: ```suggestion
This makes sudden "pop off" performances made with the No Fail mod less punishing and more rewarding, and reduces the ability for the NoFail mod to be used by unscrupulous players to "sandbag" for rank restricted tournaments. Win-win.
```

    + Old value: Aim and speed values reduced by ```0.97^x``` where x equals the number of misses.

To help illustrate these changes better, consult the graph image below. The green line refers to the old formula, and the purple line refers to the new one. You can also view this graph live and play around with the settings following [this link](https://www.desmos.com/calculator/rshij9757a).

![](/wiki/shared/news/2021-01-14-performance-points-updates/miss-penalty-curve.png)

This change seeks to penalize misses on short maps more, while significantly reducing the impact of a small number of misses on longer maps. The one-off 3% penalty for a miss provides incentive for players to hone any 1-miss plays they might still make, but otherwise makes them considerably less punishing on longer plays.

## Speed + Accuracy Curve Adjustments

As osu! has evolved over time, accuracy has slowly but steadily fallen to the wayside as a principle component for determining the pp values of modern scores. To illustrate this, 500-700pp scores from many years ago sometimes derived as much as 20-40% of their overall pp purely from the accuracy component alone.

> Today, many scores pull less than 20% of their score instead, relying on raw speed as the largest influencing component on how scores are set. This has reduced the meaningful impact of accuracy as a skill considerably, which is a peculiar approach for a rhythm game where accuracy should be king.

Reviewer 2: ```suggestion
Today, many scores pull less than 20% of their score from accuracy instead, relying on raw speed as the largest influencing component on how scores are set. This has reduced the meaningful impact of accuracy as a skill considerably, which is a peculiar approach for a rhythm game where accuracy should be king.
```

As suggested by [**Xexxar**](https://osu.ppy.sh/users/2773526), the following change has been applied:

- A new speed curve has been introduced which factors in Overall Difficulty and accuracy.
    + New value: ![](/wiki/shared/news/2021-01-14-performance-points-updates/new-speed-curve.png)

>     + Old value: ![](/wiki/shared/news/2021-01-14-performance-points-updates/old-speed-curve.png)

Reviewer 2: As a piece of commentary on this, I'd add:

> * The new speed curve, in opposition to the previous one, is no longer linear with respect to the accuracy, opting for an exponential instead, severely nerfing low-accuracy scores - especially around the 60% accuracy mark.

- Speed value is now scaled with the number of 50s made in a score in order to penalize doubletapping.
    + If there is less than one 50 per 500 objects in a map, the speed value is reduced by ```.98^x``` where x is *half the number of 50s*.

>     + If there are more than one 50 per 500 objects in a map, the speed value is reduced by ```.98^x``` where x is *the number of 50s minus the half the total number of objects in the map divided by 500*.

Reviewer 2: The OP of this pull was not updated to the latest changes; see [this comment](https://github.com/ppy/osu/pull/11175#issuecomment-744631553). This should say something like:

* If there is less than one 50 per 500 objects, the factor is a constant 1 (the speed value is unchanged; technically it's 0.98 to the zeroth power - which is 1)
* If there is more than one 50 per 500 objects, the factor is 0.98 to the power of (50 count - total count / 500).

The desmos link is wrong too, as far as I can tell. Should look something like this (the blue graph on the desmos link in the aforementioned comment looks the same):

![2021-01-12-181459_1526x674_scrot](https://user-images.githubusercontent.com/20418176/104349335-16294b80-5503-11eb-88bc-37829a36705c.png)

    + Old value: N/A

To help illustrate these changes, consult the graph images below:

> Speed curve featuring overall difficulty 8.5 & accuracy (red line is the old formula, purple line is the proposed formula, blue line is the difference curve):

Reviewer 2: ```suggestion
Speed curve featuring overall difficulty 8.5 & accuracy on the X axis (red line is the old formula, purple line is the proposed formula, blue line is the difference curve):
```

> ![](/wiki/shared/news/2021-01-14-performance-points-updates/accuracy-speed-curve-graph.png)

Reviewer 2: Values of x > 1 on this graph don't matter, but the blue graph shouldn't be cut off as it is on the negative Y axis.

[View a live version of this graph with an editable OD slider here.](https://www.desmos.com/calculator/8qfjfnto4r)

Speed value scaling with number of 50s (doubletapping penalty, x-axis is the number of 50s, y-axis is the multiplier):

![](/wiki/shared/news/2021-01-14-performance-points-updates/doubletap-penalty.png)

[View a live version of this graph with an editable total hitobject slider here.](https://www.desmos.com/calculator/tzbx9fghnl)

> All of this is a lot of math to digest, especially if you don't understand Minecraft enchanting table language that is math notation. Put shortly, these changes essentially make accuracy more of a big deal at higher levels of play and help prevent the kind of cheese we have seen in recent years with some players submitting low accuracy scores on intense maps and ending up well above their actual skill range because of it.

Reviewer 5: add a "the" between "understand" and "Minecraft" here?

---
layout: post
title: "Performance Points Updates"
date: 2021-01-14 10:00:00 +0000
---

Thanks to the effort of a number of public contributors, we're excited to announce a new set of changes to osu!'s performance points (pp) algorithm that should help freshen things up at high levels of play. Read on for more details!

![](/wiki/shared/news/2021-01-14-performance-points-updates/star-rating.jpg)

Set off during the last few days of 2020, these new incoming changes represent an effort from various members of the community to try and address long-standing issues with one of the game's most important progression systems.

For those unfamiliar with the technical minutiae of the system, please consult [the wiki article](/wiki/Performance_points) for an overview of the general ideas involved. 

To summarize the article very briefly (you should really read the whole thing on your own first!), pp encompasses several core values as follows:

> - **Aim**, which considers how difficult it is to consistently hit notes back-to-back during play (affected by things like approach rate, jumps, Flashlight, Hidden, Hard Rock and more)
> - **Speed**, which considers how quickly new elements are presented to the player during the course of the map (affected by things like the number of hit objects over time, BPM, Double Time and Half Time)
> - **Accuracy**, which considers the player's ability to stay on time rhythmically and how difficult this is to achieve (affected by things like Hidden, Hard Rock, Flashlight and more)
> - **Strain**, which considers how often and how long a player has to endure intense or technically demanding patterns during play (affected by things like streams, fast jumps and more)

Reviewer 7: ```suggestion
- **Aim**, which considers how difficult it is to consistently hit notes back-to-back during play (affected by things like approach rate, jumps, Flashlight, Hidden, Hard Rock, and more)
- **Speed**, which considers how quickly new elements are presented to the player during the course of the map (affected by things like the number of hit objects over time, BPM, Double Time, and Half Time)
- **Accuracy**, which considers the player's ability to stay on time rhythmically and how difficult this is to achieve (affected by things like Hidden, Hard Rock, Flashlight, and more)
- **Strain**, which considers how often and how long a player has to endure intensely or technically demanding patterns during play (affected by things like streams, fast jumps, and more)
```
minor formatting fixes

Reviewer 3: intense is correct; commas not being there are correct (stylistic choice, see https://en.wikipedia.org/wiki/Serial_comma)

As suggested by [**Xexxar**](https://osu.ppy.sh/users/2773526), the following change has been applied:

- The buff for AR 11 maps now scales from 0% at 0 objects, to 26.66% at a maximum of 1000 objects. The previous 20% buff is now attained at around 750 objects instead.
    - Old value: 20% flat.

In essence, this nerfs the amount of pp awarded for shorter AR 11 maps, but buffs the amount gained for longer, more difficult performances.

Consult the graph below for a visualization of the changes:

![](/wiki/shared/news/2021-01-14-performance-points-updates/ar_multiplier.png)

## No Fail Changes

Before today's changes, the performance points algorithm applied a flat 10% reduction in the maximum possible pp (not score) you could receive for a play made with the No Fail mod active.

As suggested by [**StanR**](https://osu.ppy.sh/users/7217455) in [this pull request](https://github.com/ppy/osu-performance/pull/127), the following change has been applied:

> - No Fail's pp multiplier now scales based on the amount of misses.

Reviewer 7: ```suggestion
- No Fail's pp multiplier now scales based on the number of misses.
```
the quantifier "amount" is only used with uncountable nouns

>     - New value: 1.0x base, with -0.02x applied for each miss to a maximum reduction back to 0.9x
>     - Old value: 0.9x (always)

Reviewer 7: ```suggestion
    - **New value**: 1.0x base, with -0.02x applied for each miss to a maximum reduction back to 0.9x
    - **Old value**: 0.9x (always)
```
kinda felt like these should be formatted in bold. please apply the same to the rest if you agree

The score reduction multiplier of 0.5x remains unchanged.

> This makes sudden "pop off" performances made with the No Fail mod less punishing and more rewarding, and reduces the ability for the NoFail mod to be used by unscrupulous players to "sandbag" for rank restricted tournaments. Win-win.

Reviewer 7: ```suggestion
This makes sudden "pop-off" performances made with the No Fail mod less punishing and more rewarding, and reduces the ability for the NoFail mod to be used by unscrupulous players to "sandbag" for rank restricted tournaments. Win-win.
```
iirc this needs a hyphen

## Spun Out Changes

> Similarly to how NoFail used to function as described above, Spun Out also applied a flat 5% reduction in the maximum possible pp (again, not score) you could receive while making a play with it enabled.

Reviewer 7: ```suggestion
Similar to how NoFail used to function as described above, Spun Out also applied a flat 5% reduction in the maximum possible pp (again, not score) you could receive while making a play with it enabled.
```

As suggested by [**StanR**](https://osu.ppy.sh/users/7217455) in [this pull request](https://github.com/ppy/osu-performance/pull/110), the following change has been applied:

> - Spun Out's pp multiplier now scales based on the amount of spinners in the map.

Reviewer 7: ```suggestion
- Spun Out's pp multiplier now scales based on the number of spinners in the map.
```
same as above, wrong quantifier usage

This meant that 3 misses on a 100 note map was functionally equivalent to 3 misses on a 1000 note map.

As suggested by [**Xexxar**](https://osu.ppy.sh/users/2773526), the following change has been applied:

- The miss penalty curve has been adjusted.
    - Any number of misses on a map instantly incurs a once-off 3% penalty to total pp awarded for both aim and speed.
    - The first few misses are penalised less harshly than the ones after. The largest relative punishment occurs somewhere around the 15% miss mark (relative to the number of objects in the map), after which point the reduction for each subsequent miss tapers off again.
    - Old value: Aim and speed values reduced by ```0.97^x``` where x equals the number of misses.

To help illustrate these changes better, consult the graph image below. The green line refers to the old formula, and the purple line refers to the new one. You can also view this graph live and play around with the settings following [this link](https://www.desmos.com/calculator/rshij9757a).

![](/wiki/shared/news/2021-01-14-performance-points-updates/miss-penalty-curve.png)

> This change seeks to penalize misses on short maps more, while significantly reducing the impact of a small number of misses on longer maps. The one-off 3% penalty for a miss provides incentive for players to hone any 1-miss plays they might still make, but otherwise makes them considerably less punishing on longer plays.

Reviewer 7: ```suggestion
This change seeks to penalize misses on short maps more, while significantly reducing the impact of a small number of misses on longer maps. The one-off 3% penalty for a miss provides an incentive for players to hone any 1-miss plays they might still make but otherwise makes them considerably less punishing on longer plays.
```
missing determiner and unnecessary comma usage

Reviewer 3: `an` isn't required here. comma is there for pacing. i think it's fine as-is.

## Speed & Accuracy Curve Adjustments

As osu! has evolved over time, accuracy has slowly but steadily fallen to the wayside as a principle component for determining the pp values of modern scores. To illustrate this, 500-700pp scores from many years ago sometimes derived as much as 20-40% of their overall pp purely from the accuracy component alone.

Today, many scores pull less than 20% of their score from accuracy instead, relying on raw speed as the largest influencing component on how scores are set. This has reduced the meaningful impact of accuracy as a skill considerably, which is a peculiar approach for a rhythm game where accuracy should be king.

As suggested by [**Xexxar**](https://osu.ppy.sh/users/2773526), the following change has been applied:

- A new speed curve has been introduced which factors in Overall Difficulty and accuracy.
    - New value: ![](/wiki/shared/news/2021-01-14-performance-points-updates/new-speed-curve.png)
    - Old value: ![](/wiki/shared/news/2021-01-14-performance-points-updates/old-speed-curve.png)

> - The new speed curve, in opposition to the previous one, is no longer linear with respect to accuracy, opting for an exponential instead. This severely nerfs low-accuracy scores, especially around the 60% accuracy mark.

Reviewer 7: ```suggestion
- The new speed curve, in opposition to the previous one, is no longer linear with respect to the accuracy, opting for an exponential instead. This severely nerfs low-accuracy scores, especially around the 60% accuracy mark.
```
lack of determiner

Reviewer 3: if you add "the" here you also need to make it "the accuracy of a play" or something. fine as-is.

> - Speed value is now scaled with the number of 50s made in a score in order to penalize doubletapping.

Reviewer 7: ```suggestion
- The speed value is now scaled with the number of 50s made in a score in order to penalize doubletapping.
```

    - If there is less than one 50 per 500 objects, the factor is a constant 1 (the speed value is unchanged; technically it's 0.98 to the zeroth power - which is 1)
    - If there is more than one 50 per 500 objects, the factor is 0.98 to the power of (50 count - total count / 500).
    - Old value: N/A

To help illustrate these changes, consult the graph images below:

> Speed curve featuring overall difficulty 8.5 & accuracy on the X axis (red line is the old formula, purple line is the proposed formula, blue line is the difference curve):

Reviewer 7: ```suggestion
Speed curve featuring overall difficulty 8.5 & accuracy on the X-axis (red line is the old formula, the purple line is the proposed formula, the blue line is the difference curve):
```

![](/wiki/shared/news/2021-01-14-performance-points-updates/accuracy-speed-curve-graph.png)

[View a live version of this graph with an editable OD slider here.](https://www.desmos.com/calculator/8qfjfnto4r)

> Speed value scaling with number of 50s (doubletapping penalty, x-axis is the number of 50s, y-axis is the multiplier):

Reviewer 7: ```suggestion
Speed value scaling with the number of 50s (doubletapping penalty, x-axis is the number of 50s, y-axis is the multiplier):
```

![](/wiki/shared/news/2021-01-14-performance-points-updates/doubletap-penalty.png)

[View a live version of this graph with an editable total hitobject slider here.](https://www.desmos.com/calculator/rzkcvdthtp)

All of this is a lot of math to digest, especially if you don't understand the Minecraft enchanting table language that is math notation. Put shortly, these changes essentially make accuracy more of a big deal at higher levels of play and help prevent the kind of cheese we have seen in recent years with some players submitting low accuracy scores on intense maps and ending up well above their actual skill range because of it.

---

> A big thank you to everyone involved in these changes, and especially to [**Xexxar**](https://osu.ppy.sh/users/2773526), whom has floated the major triad of the top-level changes this time around, performing all the formulaic wizardry required to make them tick largely on his own.

Reviewer 7: ```suggestion
A big thank you to everyone involved in these changes, and especially to [**Xexxar**](https://osu.ppy.sh/users/2773526), who has floated the major triad of the top-level changes this time around, performing all the formulaic wizardry required to make them tick largely on his own.
```
for being an objective pronoun, "whom" does not fit here, since you're referring to the subject
